File: tests/sample.kt
data class User(val id: Int, val name: String, val email: String)

sealed class Result<out T> {
    data class Success<T>(val data: T) : Result<T>()
    data class Error(val message: String) : Result<Nothing>()
}

class UserRepository {
    private val db = mapOf(
        1 to User(1, "Alice", "alice@example.com"),
        2 to User(2, "Bob", "bob@example.com"),
    )

    fun findById(id: Int): Result<User> =
        db[id]?.let { Result.Success(it) }
            ?: Result.Error("User $id not found")
}

fun main() {
    val repo = UserRepository()
    when (val result = repo.findById(1)) {
        is Result.Success -> println("Hello, ${result.data.name}!")
        is Result.Error   -> println("Error: ${result.message}")
    }
}

File: tests/sample.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace SpacePreview.Demo;

public record User(int Id, string Name, string Email);

public interface IUserRepository
{
    Task<User?> FindByIdAsync(int id);
    Task<IReadOnlyList<User>> ListAllAsync();
}

public sealed class InMemoryUserRepository : IUserRepository
{
    private readonly Dictionary<int, User> _store = new()
    {
        [1] = new User(1, "Alice", "alice@example.com"),
        [2] = new User(2, "Bob",   "bob@example.com"),
    };

    public Task<User?> FindByIdAsync(int id) =>
        Task.FromResult(_store.GetValueOrDefault(id));

    public Task<IReadOnlyList<User>> ListAllAsync() =>
        Task.FromResult<IReadOnlyList<User>>([.. _store.Values]);
}

class Program
{
    static async Task Main()
    {
        IUserRepository repo = new InMemoryUserRepository();
        var user = await repo.FindByIdAsync(1);
        Console.WriteLine(user is not null ? $"Hello, {user.Name}!" : "Not found.");
    }
}
